anndata/_core/sparse_dataset.py

```diff
@@ -343,7 +343,29 @@ def __repr__(self) -> str:
     def __getitem__(self, index: Index | tuple[()]) -> float | ss.spmatrix:
         row, col = self._normalize_index(index)
         mtx = self._to_backed()
-        sub = mtx[row, col]
+
+        # see https://github.com/scverse/anndata/issues/1224 for special handling of boolean masks
+        # scipy internally converts boolean masks to integer indices so this needs to occur before access to `mtx`.
+        def make_slices(mask):
+            return np.ma.extras._ezclump(mask)
+
+        def mean_slice_length(slices):
+            return floor((slices[-1].stop - slices[0].start) / len(slices))
+
+        if (
+            self.format == "csr"
+            and np.array(row).dtype == bool
+            and mean_slice_length(make_slices(row)) > 7
+        ):
+            sub = ss.vstack([mtx[s, col] for s in make_slices(row)])
```

Comment:
Instead of creating and concatenating sparse matrices, could we construct the underlying arrays directly? This could be done like `get_compressed_vectors`

Reply:
Yeah, I have definitely gotten very confused while writing sparse matrix code. Just a lot of offsets to keep in mind.